wip: yk/op-prewarm-txn

yongkangc opened the pull request:
wip

yongkangc commented:
remaining: cleanup + benching

yongkangc commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -91,24 +96,58 @@ where
+            let mut is_first_tx = true;
 
-                if handles.len() <= task_idx {
-                    let (tx, rx) = mpsc::channel();
```
thats a better way

yongkangc commented on crates/engine/tree/src/tree/payload_processor/mod.rs:
```diff
@@ -284,15 +285,16 @@ where
     ) -> CacheTaskHandle
     where
         P: BlockReader + StateProviderFactory + StateReader + Clone + 'static,
+        Tx: ExecutableTxFor<Evm> + Clone + Send + std::any::Any + 'static,
```
This is required because in prewarm.rs, we use std::any::Any for runtime type checking in the check_is_deposit_tx function. Not sure if theres a better way to go about this? 

cc @mattsse

emhane commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -27,6 +28,9 @@ use std::{
 };
 use tracing::{debug, trace};
 
+/// Optimism deposit transaction type (0x7E/126).
+const OPTIMISM_DEPOSIT_TX_TYPE: u8 = 126;
+
```
pls remove in favour of importing `DEPOSIT_TX_TYPE_ID` from op-alloy

emhane replied:
ah perhaps it makes sense to have this double up for now, since we probably don't want to have op-alloy be a dep to any crate outside of the optimism directory...

emhane commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -91,24 +99,56 @@ where
+                // For Optimism chains: The first transaction is always a deposit transaction
+                // that sets critical metadata (L1 block info, fees) affecting all subsequent
+                // transactions. We broadcast the first transaction to all workers to ensure
+                // they have this critical state.
+                if is_first_tx && executable.tx().is_type(OPTIMISM_DEPOSIT_TX_TYPE) {
+                    is_first_tx = false;
+
+                    // Pre-spawn all workers for the deposit transaction broadcast
+                    for _ in handles.len()..max_concurrency {
+                        let (tx, rx) = mpsc::channel();
+                        let sender = actions_tx.clone();
+                        let ctx = ctx.clone();
+                        let done_tx = done_tx.clone();
+
+                        executor.spawn_blocking(move || {
+                            ctx.transact_batch(rx, sender, done_tx);
+                        });
+
+                        handles.push(tx);
+                    }
 
-                    executor.spawn_blocking(move || {
-                        ctx.transact_batch(rx, sender, done_tx);
-                    });
+                    // Send deposit transaction to ALL prewarm tasks
+                    // This ensures all workers have the updated L1 block info and fee parameters
+                    for handle in &handles {
+                        let _ = handle.send(executable.clone());
+                    }
```
instead of having `is_first_tx`, it makes sense to move all this code out of the while loop and before the while loop add
```rust
if let (Ok(first_executable)) = pending.recv() {
    if executable.tx().is_type(DEPOSIT_TX_TYPE_ID) {
        ...
    }
}
```
since `is_first_tx` never resets to true, and no code is shared between the if and else bodies except the single line `executing += 1;` which is ok to repeat

mattsse commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -78,37 +86,72 @@ where
-                    executor.spawn_blocking(move || {
-                        ctx.transact_batch(rx, sender, done_tx);
-                    });
+            // Spawn the required number of workers
+            for _ in 0..workers_needed {
```
now that I see this, do we even configure the max concurrency anywhere, because spawning 64 tasks is probably not ideal

yongkangc replied:
i think its mostly hardcoded here as 64 here
https://github.com/paradigmxyz/reth/blob/yk/op-txn-prewarming/crates/engine/tree/src/tree/payload_processor/prewarm.rs#L89

Given that a normal txns prob average around 150 (mainnet) https://dune.com/queries/3149880 and base around 250 https://base.blockscout.com/blocks

16 would probably work?

mattsse commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -78,37 +86,72 @@ where
-                    handles.push(tx);
+            // Handle first transaction - special case for Optimism deposit / systems txs
+            if let Ok(first_tx) = pending.recv() {
+                if first_tx.tx().is_type(OPTIMISM_DEPOSIT_TX_TYPE) {
```
we can do > 4 here because this will catch all non eth types and if its first position we can assume this is a sys tx

yongkangc replied:
this makes sense, thanks for the note!

  - Type 0: Legacy transactions (original Ethereum transactions)
  - Type 1: EIP-2930 transactions (with access lists)
  - Type 2: EIP-1559 transactions (with base fee + priority fee)
  - Type 3: EIP-4844 transactions (blob transactions for L2 data)
  - Type 4: EIP 7702

mattsse commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -25,7 +26,28 @@ use std::{
+///
+/// Any transaction with a type > 4 is considered a non-standard/system transaction,
+/// typically used by L2s for special purposes (e.g., Optimism deposit transactions use type 126).
+const MAX_STANDARD_TX_TYPE: u8 = 4;
```
we're basically optimistically optimizing for a specific case where the first tx in a block is different, which indicates a system tx (op,arb,bera,..) which should be okay

yongkangc commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -78,37 +124,74 @@ where
+                    // Broadcast system transaction to all workers to ensure they have the
+                    // critical state. This is particularly important for L2s like Optimism
+                    // where the first deposit transaction contains essential block metadata.
+                    for handle in &handles {
+                        if handle.send(first_tx.clone()).is_err() {
```
this makes sense, technically we should preskip x txns, where x >1

previously tried experiments here
https://github.com/paradigmxyz/reth/pull/18122/files

yongkangc commented on crates/engine/tree/src/tree/payload_processor/prewarm.rs:
```diff
@@ -92,38 +128,90 @@ where
-                    executor.spawn_blocking(move || {
-                        ctx.transact_batch(rx, sender, done_tx);
-                    });
+            let workers_needed = transaction_count_hint.max(1).min(max_concurrency);
```
> we can do both limit the default to 16 but also check the tx_hint upper bound 

not sure if this is what you had in mind for the settings? @mattsse

yongkangc commented:
@mattsse thanks for the comments, addressed them - added in a better way to limit workers based on type hint and 16 + added in tx index 

not 100% sure if thats what u mean with the index?

mediocregopher reviewed (approved):
This LGTM but maybe @mattsse should double check I'm not missing something

yongkangc commented:
<img width="1018" height="647" alt="image" src="https://github.com/user-attachments/assets/4b2f80b0-b417-4aac-a6da-d06280a6279c" />
@mattsse benched and got this for reth7

yongkangc commented:
@mattsse just addressed this
https://github.com/paradigmxyz/reth/pull/18327#discussion_r2387015130

here: https://github.com/paradigmxyz/reth/pull/18327/commits/26ad92033eb6315729549536a62d3da367019892

let me know if this addresses your comment
